File: Collections/Queue.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace Collections
{
    /// <summary>
    /// Linked list implementation of a queue.
    /// </summary>
    public class Queue<T> : IQueue<T>
    {
        private Node head;  // refers to the first element of the queue
        private Node tail;  // refers to the last element of the queue
        private int size = 0;

        private class Node
        {
            public T payload;
            public Node next;
        }

        public Queue(params T[] elements)
        {
            foreach (var element in elements)
                Enqueue(element);
        }

        public T Dequeue()
        {
            if (IsEmpty())
            {
                throw new InvalidOperationException();
            }
            var payload = head.payload;
            head = head.next;
            size--;
            return payload;
        }

        public void Enqueue(T payload)
        {
            var last = new Node { payload = payload };
            if (IsEmpty())
            {
                head = last;
                tail = last;
            }
            else
            {
                tail.next = last;
                tail = last;
            }
            size++;
        }

        public T Peek()
        {
            if (IsEmpty())
            {
                throw new InvalidOperationException();
            }
            return head.payload;
        }

        public void Push(T payload)
        {
            Enqueue(payload);
        }

        public T Pop() => Dequeue();

        public int Size() => size;

        public bool IsEmpty() => size == 0;

        public IEnumerator<T> GetEnumerator()
        {
            var current = head;
            while (current != tail)
            {
                yield return current.payload;
                current = current.next;
            }
        }

        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    }
}

File: src/main/java/in/bfdes/collections/Queue.java
package in.bfdes.collections;

import java.util.Iterator;
import java.util.NoSuchElementException;

public class Queue<T> implements Iterable<T> {
    private Node head;
    private Node tail;
    private int size = 0;

    private class Node {
        public final T item;
        public Node next;

        public Node(T item, Node next) {
            this.item = item;
            this.next = next;
        }
    }

    public Queue() {
    }

    public Queue(T[] initialItems) {
        for (var item : initialItems)
            push(item);
    }

    public Queue(Iterable<T> initialItems) {
        if (initialItems == null)
            throw new IllegalArgumentException();
        for (var item : initialItems)
            push(item);
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public T peek() {
        if (isEmpty())
            throw new NoSuchElementException();
        return head.item;
    }

    public T pop() {
        if (isEmpty())
            throw new NoSuchElementException();
        var item = head.item;
        head = head.next;
        size--;
        return item;
    }

    public void push(T item) {
        if (item == null)
            throw new IllegalArgumentException();
        var last = new Node(item, null);
        if (isEmpty()) {
            head = last;
        } else {
            tail.next = last;
        }
        tail = last;
        size++;
    }

    public T dequeue() {
        return pop();
    }

    public void enqueue(T item) {
        push(item);
    }

    @Override
    public Iterator<T> iterator() {
        return new Iterator<>() {
            private Node next = head;

            public boolean hasNext() {
                return next != null;
            }

            public T next() {
                if (!hasNext())
                    throw new NoSuchElementException();
                T item = next.item;
                next = next.next;
                return item;
            }
        };
    }
}
